File: Python/Arrays/L_largestSubMatWRearr.py
from typing import List


def largestSubmatrix(matrix: List[List[int]]) -> int:
    row, col = len(matrix), len(matrix[0])

    # Calculate heights for each column
    for i in range(1, row):
        for j in range(col):
            if matrix[i][j] == 1:
                matrix[i][j] += matrix[i - 1][j]

    res = 0
    for i in range(row):
        # Sort the heights in ascending order
        matrix[i].sort()

        # Iterate through the sorted heights
        for j in range(col):
            height = matrix[i][j]
            width = col - j
            res = max(res, height * width)

    return res

File: cpp/Array/L_largestSubMatWRearr.cpp
#include <bits/stdc++.h>
using namespace std;

int largestSubmatrix(vector<vector<int>> &matrix)
{
    int row = matrix.size();
    int col = matrix[0].size();

    // Calculate heights for each column
    for (int i = 1; i < row; i++)
    {
        for (int j = 0; j < col; j++)
        {
            if (matrix[i][j] == 1)
            {
                matrix[i][j] += matrix[i - 1][j];
            }
        }
    }

    int res = 0;
    for (int i = 0; i < row; i++)
    {
        // Sort the heights in ascending order
        sort(matrix[i].begin(), matrix[i].end());

        // Iterate through the sorted heights
        for (int j = 0; j < col; j++)
        {
            int height = matrix[i][j];
            int width = col - j;
            res = max(res, height * width);
        }
    }

    return res;
}